algo/src/main/kotlin/com/hartwig/actin/algo/evaluation/treatment/HasHadTreatmentWithDrugWithCycles.kt

```diff
@@ -0,0 +1,73 @@
+package com.hartwig.actin.algo.evaluation.treatment
+
+import com.hartwig.actin.algo.evaluation.EvaluationFactory
+import com.hartwig.actin.algo.evaluation.EvaluationFunction
+import com.hartwig.actin.algo.evaluation.treatment.MedicationFunctions.createTreatmentHistoryEntriesFromMedications
+import com.hartwig.actin.algo.evaluation.util.Format.concatItemsWithAnd
+import com.hartwig.actin.algo.evaluation.util.Format.concatItemsWithOr
+import com.hartwig.actin.datamodel.PatientRecord
+import com.hartwig.actin.datamodel.algo.Evaluation
+import com.hartwig.actin.datamodel.algo.EvaluationResult
+import com.hartwig.actin.datamodel.clinical.treatment.Drug
+import com.hartwig.actin.datamodel.clinical.treatment.DrugTreatment
+
+class HasHadTreatmentWithDrugWithCycles(private val drugsToFind: Set<Drug>, private val minCycles: Int?) : EvaluationFunction {
+
+    override fun evaluate(record: PatientRecord): Evaluation {
+        val effectiveTreatmentHistory = record.oncologicalHistory + createTreatmentHistoryEntriesFromMedications(record.medications)
+        val namesToMatch = drugsToFind.map { it.name.lowercase() }.toSet()
+        val drugList = concatItemsWithOr(drugsToFind)
+
+        val drugsByEvaluationResult: Map<EvaluationResult, Set<Drug>> = effectiveTreatmentHistory
+            .flatMap { entry ->
+                val result = if (minCycles == null) EvaluationResult.PASS else when (entry.treatmentHistoryDetails?.cycles) {
+                    null -> EvaluationResult.UNDETERMINED
+                    in minCycles..Int.MAX_VALUE -> EvaluationResult.PASS
+                    else -> EvaluationResult.FAIL
+                }
```

Comment:
The name could be better, and I might have used a more general `when` to handle all cases:
```suggestion
                val hasCycles = entry.treatmentHistoryDetails?.cycles.let { cycles ->
                    when {
                        minCycles == null -> EvaluationResult.PASS
                        cycles == null -> EvaluationResult.UNDETERMINED
                        cycles >= minCycles -> EvaluationResult.PASS
                        else -> EvaluationResult.FAIL
                    }
                }
```